### ISSUE TITLE
New Chrome/Safari autoplay policy does not like ICE Lite

### ISSUE THREAD

**Issue by ibc** [versatica/mediasoup#264]
Scenario (at least in Canary 73 OSX):

* Have mediasoup-client connect to a room as viewer (no `getUserMedia()` called).
* Receive some audio and video consumers, create MediaStreams and attach them to audio and/or video HTML elements via `elem.srcObject = stream`. They can have `autoplay` attribute, or you may with to call `play()` via JS.
* You get no audio and blank video.

It does NOT happen with other WebRTC servers (such as Janus) that are full ICE, so being ICE-Lite does not help here.

Workaround given by Tim Panton:

```js
function aplay(video) {
  var promise = http://video.play ();
  if (promise !== undefined) {
    promise.then(_ => {
        console.log("autoplay worked");
        // Autoplay started!
  }).catch(error => {
    // Autoplay was prevented, mute video and play.
    video.muted = true;
    console.log("mute and play");
    http://video.play ();
  });
 }
}
```

> You'll get video, then you need a user action to unmute. After the user unmutes a few times google will allow auto play.

So, should we add full ICE support? We may just generate ICE binding requests in response to received  ICE binding requests... but then we need to handle the peer ICE uFrag and pwd... buff...

**Comment by fippo** [comment]
you have tested a completly fresh profile to avoid the MEI AI?

**Comment by ibc** [comment]
> you have tested a completly fresh profile to avoid the MEI AI?

No, I just reseted mic/webcam permission to "Ask".

**Comment by ibc** [comment]
Good news. **It's NOT ICE Lite related**

I just called `video.play()` manually while the video was blank and then got video rendering. The demo app does not even call `video.play()` but just relies on the `autoplay=true` attribute in the HTML video element.

**Comment by fippo** [comment]
that probably explains why it worked in 71 but not in 73. Debuggability is overrated :joy:

**Comment by ibc** [comment]
Comment by Pavel:

> That seems to be an unnecessary hack. You can leverage web audio api to check if autoplay prevention has been enforced:
>
```js
var ctx = new AudioContext();
console.log(ctx.state); // should be 'running', otherwise user action is required.
```
> after user action, you can check if browser is happy with it:
```js
ctx.resume().then(console.log('you're good to go));
```

**Comment by fippo** [comment]
This is quite a good hack, thank you Pavel

**Comment by vpavlov80** [comment]
I did a lot of tests for this and indeed the problem is the Chrome 71 (at least on Windows) autoplay policy. Basically you can not autostart an audio or video tag if there is no an user interaction on the page before that or if your site has a good record in Media Engagement Index (MEI) (see https://developers.google.com/web/updates/2017/09/autoplay-policy-changes). Still if the video tag is muted it can be autostarted without an user interaction.

As for the Pavel's hack it is working great for Chrome, but on FF I always get ctx.state = 'suspended', so I'm using it only for Chrome.

It would be great if mediasoup-client can help with this. It will probably save a lot of time to other developers, as debugging this issue is a nightmare especially with the MEI :). Here is my suggestion:
- when the first consumer start to receive - the mediasoup-client can do the AudioContext test;
- if the test fails - it can throw an exception, or at least show a warning.
- basically after we know that auto start doesn't work - the developer should just show a warning with a button on the page. When the user click the button - all that should be done is to go through all running consumers and add / replace their tracks in the media elements.

**Comment by vpavlov80** [comment]
Thank you,
I get what you are saying :). The steps above can be easily implemented in any code that uses mediasoup-client and they have fixed the issue for me.

**Comment by AJLoveChina** [comment]
@ibc 

It seems not working on firefox, but is OK on Chrome.

Example here : https://jsbin.com/kipacok/edit?html,console,output

![image](https://user-images.githubusercontent.com/11222746/84110810-28687600-aa58-11ea-8965-3cde17b7fd0c.png)
